pkg/mount/dag_steps.go

```diff
@@ -808,3 +788,96 @@ func (s *State) MountLiveCd(g *herd.Graph, opts ...herd.OpOption) error {
 		return nil
 	}))...)
 }
+
+// UKIBootInitDagStep tries to launch /sbin/init in root and pass over the system
+// booting to the real init process
+// Drops to emergency if not able to. Panic if it cant even launch emergency.
+func (s *State) UKIBootInitDagStep(g *herd.Graph) error {
+	return g.Add(cnst.OpUkiInit,
+		herd.WeakDeps,
+		herd.WithWeakDeps(cnst.OpRemountRootRO, cnst.OpRootfsHook, cnst.OpInitramfsHook, cnst.OpWriteFstab),
+		herd.WithCallback(func(ctx context.Context) error {
+			output, err := internalUtils.CommandWithPath("/usr/lib/systemd/systemd-pcrphase --graceful leave-initrd")
+			if err != nil {
+				internalUtils.Log.Err(err).Msg("running systemd-pcrphase")
+				internalUtils.Log.Debug().Str("out", output).Msg("systemd-pcrphase leave-initrd")
+			}
+			// Print dag before exit, otherwise its never printed as we never exit the program
+			internalUtils.Log.Info().Msg(s.WriteDAG(g))
+
+			// Mount a tmpfs under sysroot
+			err = syscall.Mount("tmpfs", s.path("sysroot"), "tmpfs", syscall.MS_NOSUID|syscall.MS_NODEV|syscall.MS_NOEXEC, "")
+
+			// Move all the dirs in root FS that are not a mountpoint to the new root via Bind mount
+			err = filepath.WalkDir(s.Rootdir, func(path string, d fs.DirEntry, err error) error {
+				if err != nil {
+					return err
+				}
+				if !d.IsDir() {
+					return nil
+				}
+				fileInfo, err := os.Stat(s.path(path))
+				if err != nil {
+					return err
+				}
+				parentPath := filepath.Dir(s.path(path))
+				parentInfo, err := os.Stat(parentPath)
+				if err != nil {
+					return err
+				}
+				// If the directory has the same device as its parent, it's not a mount point.
+				if fileInfo.Sys().(*syscall.Stat_t).Dev == parentInfo.Sys().(*syscall.Stat_t).Dev {
+					internalUtils.Log.Info().Msg(fmt.Sprintf("%s is a simple directory", path))
+					err = os.MkdirAll(filepath.Join(s.path("sysroot"), path), 0755)
```

Comment:
<img width="939" alt="image" src="https://github.com/kairos-io/immucore/assets/2794419/d2a8801e-8624-42ac-a2f6-775bff9a2ca6">

Reply:
first problem, it should never go recursively (only stick to `/dev`). Second, it won't be able to create directories anyway (?).